### ISSUE TITLE
Default permission on permission nodes

### ISSUE THREAD

**Issue by ljacqu** [Xephi/AuthMeReloaded#606]
Currently we define a default permission (nobody/OP/all) on CommandDescription to fall back to if no permission system is available.

Should it be on the permission node enum instead? We use permission nodes not only for validating command access and structurally doesn't make sense to be in the command part.

**Comment by timvisee** [comment]
I've developed the command handling and permissions part.

General permissions (non-command permissions) should be checked through the permissions manager. 

For commands, you could do the same. That would require you to write a lot of boilerplate code, the command manager integrates the permission check so you don't have to do it manually. The command manager also allows players to see whether they have permission to run a command, and it allows to only show commands to a user he has permission for. These things wouldn't be possible when it was implement in another way, I think at least.

Does this 'solve' the issue, or are you talking about something else?

**Comment by ljacqu** [comment]
@Gnat008

In `CommandInitializer` we build `CommandDescription` objects. Those contain all the metadata of our commands (how can I access it? how many arguments can it have?) including what permission is necessary in order to execute the command — you'll find a permissions field.

This permissions field holds a `CommandPermissions` class. It's not much more than a container with a list of permission nodes and a default permission. A default permission is a fallback when permission can't be checked (no permission system available). 

What's a little odd is that we unite a default permission with permission nodes only in the command part. I would expect that the default permission resides on the permission node itself: e.g. `PlayerPermission.LOGOUT` would have an additional field of `DefaultPermission` type. Then we have a default permission associated with all permission nodes and don't need to manage this in the command section of AuthMe.

I'm making an assumption here: one is that we always only have one permission node in the list of `CommandPermissions`, which I'm almost certain is the case. 

If the permission node classes contain a default permission, `CommandPermissions` can be deleted and the field in `CommandDescription` changed to a `PermissionNode` field. 

@timvisee Does this clarify the issue for you as well?

**Comment by timvisee** [comment]
@ljacqu Yes, I think it does.

To be honest, I don't know about the `CommandInitializer` and `PlayerPermission` classes. That has been done by someone else, after my work, haha.

Implementing the default permissions in the permission enum's themselve sound logical. Thus I would fully agree with that improvement.

About the last part. Do you mean that you'd like to replace the `CommandPermissions` attribute in the command object with an attribute holding a permission enum? If so, then yes.

One thing to note, the command permissions class supports multiple permission nodes to be set for a command. This feature will be removed with this change (unless you'd implement the attribute as a collection of some sort). If this feature isn't used (and I don't think it is), it wouldn't be a problem.

**Comment by ljacqu** [comment]
@timvisee Thanks for the feedback!

To the points you mention – I created the `CommandInitializer` (it's a separation of the command handling class, again an issue of a class worrying about too many problems). It just builds all of the `CommandDescription` objects. I created a Wiki page about our [command handling](https://github.com/AuthMe-Team/AuthMeReloaded/wiki/Command-Handling), though the constructor graph is outdated.

Good point about the new restriction, only one permission node per command. That's an intentional restriction; currently we don't use it for any command so this translates for me to unused code that needs to be maintained (I wager there's not even a unit test covering that scenario).
